docs/CONTEXT.md

```diff
@@ -0,0 +1,96 @@
+# Context object
+`HullApp` applies multiple middleware to the request handler.
+The result is `req.hull` object which is the Context Object - a set of modules to work in the context of current organization and connector instance.
+
+```javascript
+{
+  client: {
+    logger: {},
+  },
+  ship: {},
+  token: "",
+  segments: [],
+  hostname: req.hostname,
+  cache: {},
+  enqueue: () => {},
+  metric: {},
+  service: {},
+  message: {},
+  notification: {}
+}
+```
+
+#### client
+[Hull API client](../README.md) initialized to work with current organization
+
+#### ship
+Ship object with manifest information and `private_settings`
+
+#### token
+A string with encoded credentials
+
+#### segments
+An array of segments information
+
+#### hostname
+Hostname of the current request.
+
+#### cache
+[Caching object](../src/infra/cache/ship-cache.js)
+
+#### enqueue
+A [function](../src/infra/queue/enqueue.js) to perform tasks in async manner.
+
+#### metric
+An [object](../src/infra/instrumentation/metric-agent.js) to send data to metrics service.
+
+#### service
+A namespace reserved for connector developer to inject a custom logic, which needs to work in the context.
```

Comment:
Sure, I will add more docs.
The basic idea is to structure the situation like this:
https://github.com/hull-ships/hull-stripe/blob/master/server/actions/fetch-history.js

where you have multiple functions: `fetchEventPage`, `storeEvent`, `fetchUserPage`, `storeUser`, which all needs `req.hull.client` to operate. What if you would like to add metrics instrumentation there? And then ship instance to get some settings to customize the behaviour? It could lead to altering the function signatures to pass next two dependencies:
`fetchEventPage({ customers, hull, stripe, cursor, metric, ship })`, `fetchUserPage({ hull, stripe, cursor, metric, ship })`, `fetchHistory({ clientSecret, hull, metric, ship })`.

With the service namespace and the context binding you can ideally have:

`fetchHistory(ctx, { clientSecret })` and `fetchEventPage(ctx, { customers, cursor })` and `fetchUserPage(ctx, { stripe, cursor })` and then you can use them within the route:

```js
app.post("/fetch-all", ...middlewares, (req, res) => {
  fetchHistory(req.hull, { clientSecret })
  .then(() => {}, () => {});
});
```